File: Core/Model/Impl/Switch.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Xml.Serialization;
using BinkyRailways.Core.Util;

namespace BinkyRailways.Core.Model.Impl
{
    /// <summary>
    /// Standard two-way left/right switch
    /// </summary>
    public sealed class Switch : Junction, ISwitch
    {
        private readonly Property<bool> hasFeedback;
        private readonly Property<int> switchDuration;
        private readonly Property<bool> invert;
        private readonly Property<bool> invertFeedback;
        private readonly Property<SwitchDirection> initialDirection;

        /// <summary>
        /// Default ctor
        /// </summary>
        public Switch()
            : base(16, 12)
        {
            hasFeedback = new Property<bool>(this, true);    
            switchDuration = new Property<int>(this, 1000);
            invert = new Property<bool>(this, false);
            invertFeedback = new Property<bool>(this, false);
            initialDirection = new Property<SwitchDirection>(this, SwitchDirection.Straight);
        }

        /// <summary>
        /// Address used to control the switch
        /// </summary>
        [XmlIgnore]
        public Address Address { get; set; }

        /// <summary>
        /// Address of the feedback unit of the entity
        /// </summary>
        [XmlIgnore]
        public Address FeedbackAddress { get; set; }

        /// <summary>
        /// Gets all (non-null) addresses configured in this entity.
        /// </summary>
        [XmlIgnore]
        public IEnumerable<AddressUsage> AddressUsages
        {
            get
            {
                var addr = Address;
                if (addr != null) yield return new AddressUsage(addr, AddressDirection.Output);
                if (HasFeedback)
                {
                    addr = FeedbackAddress ?? Address;
                    if (addr != null) yield return new AddressUsage(addr, AddressDirection.Input);
                }
            }
        }

        /// <summary>
        /// Does this switch send a feedback when switched?
        /// </summary>
        public bool HasFeedback
        {
            get { return hasFeedback.Value; }
            set { hasFeedback.Value = value; }
        }

        /// <summary>
        /// Time (in ms) it takes for the switch to move from one direction to the other?
        /// This property is only used when <see cref="HasFeedback"/> is false.
        /// </summary>
        public int SwitchDuration
        {
            get { return switchDuration.Value; }
            set { switchDuration.Value = Math.Max(0, value); }
        }

        /// <summary>
        /// If set, the straight/off commands are inverted.
        /// </summary>
        public bool Invert
        {
            get { return invert.Value; }
            set { invert.Value = value; }
        }

        /// <summary>
        /// If there is a different feedback address and this is set, the straight/off feedback states are inverted.
        /// </summary>
        public bool InvertFeedback
        {
            get { return invertFeedback.Value; }
            set { invertFeedback.Value = value; }
        }

        /// <summary>
        /// At which direction should the switch be initialized?
        /// </summary>
        [DefaultValue(SwitchDirection.Straight)]
        public SwitchDirection InitialDirection
        {
            get { return initialDirection.Value; }
            set { initialDirection.Value = value; }
        }

        /// <summary>
        /// Address of the locomotive
        /// </summary>
        [XmlElement("Address")]
        [Browsable(false)]
        [EditorBrowsable(EditorBrowsableState.Never)]
        public string AddressString
        {
            get { return Default<AddressTypeConverter>.Instance.ConvertToString(Address); }
            set { Address = (Address)Default<AddressTypeConverter>.Instance.ConvertFromString(value); }
        }

        /// <summary>
        /// Address of the feedback unit
        /// </summary>
        [XmlElement("FeedbackAddress")]
        [Browsable(false)]
        [EditorBrowsable(EditorBrowsableState.Never)]
        public string FeedbackAddressString
        {
            get { return Default<AddressTypeConverter>.Instance.ConvertToString(FeedbackAddress); }
            set { FeedbackAddress = (Address)Default<AddressTypeConverter>.Instance.ConvertFromString(value); }
        }

        /// <summary>
        /// Accept a visit by the given visitor
        /// </summary>
        public override TReturn Accept<TReturn, TData>(EntityVisitor<TReturn, TData> visitor, TData data)
        {
            return visitor.Visit(this, data);
        }

        /// <summary>
        /// Validate the integrity of this entity.
        /// </summary>
        public override void Validate(IEntity validationRoot, ValidationResults results)
        {
            base.Validate(validationRoot, results);
            if (Address == null)
            {
                results.Warn(this, Strings.WarnNoAddressSpecified);
            }
            if (SwitchDuration > 5000)
            {
                results.Warn(this, Strings.WarnSwitchDurationVeryHigh);
            }
        }

        /// <summary>
        /// Human readable name of this type of entity.
        /// </summary>
        public override string TypeName
        {
            get { return Strings.TypeNameSwitch; }
        }
    }
}

File: pkg/core/model/impl/switch.go
// Copyright 2020 Ewout Prangsma
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author Ewout Prangsma
//

package impl

import (
	"context"
	"iter"

	"github.com/binkyrailways/BinkyRailways/pkg/core/model"
	"github.com/binkyrailways/BinkyRailways/pkg/core/model/refs"
)

// Switch adds implementation methods to model.Switch
type Switch interface {
	ModuleEntity
	model.Switch
}

type stdSwitch struct {
	junction

	Address          model.Address          `xml:"Address"`
	FeedbackAddress  model.Address          `xml:"FeedbackAddress,omitempty"`
	HasFeedback      *bool                  `xml:"HasFeedback,omitempty"`
	IsLeft           *bool                  `xml:"IsLeft,omitempty"`
	SwitchDuration   *int                   `xml:"SwitchDuration,omitempty"`
	Invert           *bool                  `xml:"Invert,omitempty"`
	InvertFeedback   *bool                  `xml:"InvertFeedback,omitempty"`
	InitialDirection *model.SwitchDirection `xml:"InitialDirection,omitempty"`
}

var _ Switch = &stdSwitch{}

func newSwitch() *stdSwitch {
	sw := &stdSwitch{}
	sw.EnsureID()
	sw.SetDescription("New switch")
	sw.junction.Initialize(16, 12)
	return sw
}

// GetEntityType returns the type of this entity
func (sw *stdSwitch) GetEntityType() string {
	return TypeSwitch
}

// Accept a visit by the given visitor
func (sw *stdSwitch) Accept(v model.EntityVisitor) interface{} {
	return v.VisitSwitch(sw)
}

// ForEachAddress iterates all addresses in this entity and any child entities.
func (sw *stdSwitch) ForEachAddress(cb func(addr model.Address, onUpdate func(context.Context, model.Address) error)) {
	cb(sw.Address, sw.SetAddress)
	cb(sw.FeedbackAddress, sw.SetFeedbackAddress)
}

// Get the Address of the entity
func (sw *stdSwitch) GetAddress() model.Address {
	return sw.Address
}

// Set the Address of the entity
func (sw *stdSwitch) SetAddress(ctx context.Context, value model.Address) error {
	if !sw.Address.Equals(value) {
		if sw.GetDescription() == sw.Address.Value {
			sw.Description = value.Value
		}
		sw.Address = value
		sw.OnModified()
	}
	return nil
}

// Return a sequence of all (non-empty) addresses configured in this
// entity with the direction their being used.
func (sw *stdSwitch) AllAddressUsages() iter.Seq[model.AddressUsage] {
	return func(yield func(model.AddressUsage) bool) {
		if !sw.Address.IsEmpty() {
			if !yield(model.AddressUsage{
				Address:   sw.Address,
				Direction: model.AddressDirectionOutput,
			}) {
				return
			}
		}
		if !sw.FeedbackAddress.IsEmpty() {
			if !yield(model.AddressUsage{
				Address:   sw.FeedbackAddress,
				Direction: model.AddressDirectionInput,
			}) {
				return
			}
		}
	}
}

// Is this switch turning to the left?
// Otherwise it is turning to the right.
func (sw *stdSwitch) GetIsLeft() bool {
	return refs.BoolValue(sw.IsLeft, false)
}
func (sw *stdSwitch) SetIsLeft(value bool) error {
	if sw.GetIsLeft() != value {
		sw.IsLeft = refs.NewBool(value)
		sw.OnModified()
	}
	return nil

}

// Does this switch send a feedback when switched?
func (sw *stdSwitch) GetHasFeedback() bool {
	return refs.BoolValue(sw.HasFeedback, true)
}
func (sw *stdSwitch) SetHasFeedback(value bool) error {
	if sw.GetHasFeedback() != value {
		sw.HasFeedback = refs.NewBool(value)
		sw.OnModified()
	}
	return nil
}

// Address of the feedback unit of the entity
func (sw *stdSwitch) GetFeedbackAddress() model.Address {
	return sw.FeedbackAddress
}
func (sw *stdSwitch) SetFeedbackAddress(ctx context.Context, value model.Address) error {
	if !sw.FeedbackAddress.Equals(value) {
		sw.FeedbackAddress = value
		sw.OnModified()
	}
	return nil
}

// Time (in ms) it takes for the switch to move from one direction to the other?
// This property is only used when <see cref="HasFeedback"/> is false.
func (sw *stdSwitch) GetSwitchDuration() int {
	return refs.IntValue(sw.SwitchDuration, 1000)
}
func (sw *stdSwitch) SetSwitchDuration(value int) error {
	if sw.GetSwitchDuration() != value {
		sw.SwitchDuration = refs.NewInt(value)
		sw.OnModified()
	}
	return nil
}

// If set, the straight/off commands are inverted.
func (sw *stdSwitch) GetInvert() bool {
	return refs.BoolValue(sw.Invert, false)
}
func (sw *stdSwitch) SetInvert(value bool) error {
	if sw.GetInvert() != value {
		sw.Invert = refs.NewBool(value)
		sw.OnModified()
	}
	return nil
}

// If there is a different feedback address and this is set, the straight/off feedback states are inverted.
func (sw *stdSwitch) GetInvertFeedback() bool {
	return refs.BoolValue(sw.InvertFeedback, false)
}
func (sw *stdSwitch) SetInvertFeedback(value bool) error {
	if sw.GetInvertFeedback() != value {
		sw.InvertFeedback = refs.NewBool(value)
		sw.OnModified()
	}
	return nil
}

// At which direction should the switch be initialized?
func (sw *stdSwitch) GetInitialDirection() model.SwitchDirection {
	return refs.SwitchDirectionValue(sw.InitialDirection, model.SwitchDirectionStraight)
}
func (sw *stdSwitch) SetInitialDirection(value model.SwitchDirection) error {
	if sw.GetInitialDirection() != value {
		sw.InitialDirection = refs.NewSwitchDirection(value)
		sw.OnModified()
	}
	return nil
}
